### PR TITLE
Mask load/save: Support numpy and edf format and vertically flipped mask

### PR DESCRIPTION
**Pull request by t20100**
This PR aims at making the mask interoperable with other tools:
- Adding support for load/save mask from/to npy and edf.
- Saving the mask not flipped with respect to original data (and also allowing to load it), by providing extra options in the dialog:
  - Mask (*.mask): Unchanged behavior
  - Vertically flipped mask (*.npy *.edf): ~~Reflecting image flip applied on load to mask load/save~~ Flip mask vertically on load/save

I also started looking at a "vertically flipped (*.mask)" option: There is a TIFF "Orientation" tag to store this information. `Pillow` is interpreting it and returns the flipped array.. so it does not really help. I have left this aside for now.

~~Regarding the vertical flip, it depends on the image file loader that was used: It is only applied when loaded with PIL, fabio or lambda loader (and not with spe, karabo and h5). So since the goal is to have a mask not flipped with respect to the original data, I store the flip information in the image model and use it to flip or not the mask for load/save from the `MaskController`. This made this PR bigger than first expected.~~

I haven't yet looked at the batch yet, and since it calls `load_mask`, it will need to be adapted as well (so I leave this PR as Draft).

### CODE REVIEW

File: dioptas/model/ImgModel.py [hunk 1]
```diff
@@ -71,6 +71,7 @@ def __init__(self):
         self._img_data_background_subtracted = None
         self._img_data_absorption_corrected = None
         self._img_data_background_subtracted_absorption_corrected = None
+        self.img_flipud = False
```
Comment (t20100) [inline dioptas/model/ImgModel.py:74]: Stored by loader to be retrieved later

File: dioptas/model/MaskModel.py [hunk 2]
```diff
@@ -231,24 +232,44 @@ def set_mask(self, mask_data):
         self.update_deque()
         self._mask_data = mask_data
 
-    def save_mask(self, filename):
+    def save_mask(self, filename, flipud: bool=False):
```
Comment (t20100) [inline dioptas/model/MaskModel.py:235]: Adding a `flipud` argument so the mask model does not need access to the image model.

File: dioptas/model/MaskModel.py [hunk 3]
```diff
@@ -231,24 +232,44 @@ def set_mask(self, mask_data):
...
-            im.save(filename, "tiff", compression="tiff_deflate")
-        except OSError:
+        if flipud:
+            im_array = np.flipud(im_array)
+
+        if filename.endswith('.npy'):
+            np.save(filename, im_array)
+        elif filename.endswith('.edf'):
+            fabio.edfimage.EdfImage(im_array).write(filename)
+        else:
```
Comment (t20100) [inline dioptas/model/MaskModel.py:244]: I put new formats first, so it falls back to the previous implementation

File: dioptas/model/MaskModel.py [hunk 4]
```diff
@@ -231,24 +232,44 @@ def set_mask(self, mask_data):
...
+                    im.save(filename, "tiff")
 
         self.filename = filename
 
-    def load_mask(self, filename):
-        try:
-            data = np.array(Image.open(filename))
-        except IOError:
-            data = np.loadtxt(filename)
+    def _load_mask(self, filename, flipud: bool):
```
Comment (t20100) [inline dioptas/model/MaskModel.py:256]: Sharing loading implementation between `load_mask` and `add_mask`

File: dioptas/tests/controller_tests/test_MaskController.py [hunk 5]
```diff
@@ -57,11 +61,14 @@ def test_loading_and_saving_mask_files(self):
         QtWidgets.QFileDialog.getOpenFileName = MagicMock(return_value=os.path.join(unittest_data_path, 'test.mask'))
         click_button(self.mask_widget.load_mask_btn)
         self.model.mask_model.mask_below_threshold(self.model.img_data, 1)
-        filename = os.path.join(unittest_data_path, 'dummy.mask')
 
-        QtWidgets.QFileDialog.getSaveFileName = MagicMock(return_value=filename)
-        click_button(self.mask_widget.save_mask_btn)
-        self.assertTrue(os.path.exists(filename))
+        for extension in self.MASK_EXTENSIONS:
```
Comment (t20100) [inline dioptas/tests/controller_tests/test_MaskController.py:65]: I adapted existing to also test new mask extensions

File: dioptas/tests/functional_tests/test_mask_save_and_load.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+import numpy as np
+import pytest
+from PIL import Image
+from qtpy import QtCore, QtWidgets
+from qtpy.QtTest import QTest
+
+from ..utility import click_button, unittest_data_path, QtTest
+from ...controller.MainController import MainController
+
+
```
Comment (t20100) [inline dioptas/tests/functional_tests/test_mask_save_and_load.py:34]: I made this new test with pytest fixtures. If you prefer that it's use `QtTest`, I'll change this.

**Comment by CPrescher** [comment]
Hi, this looks good to me. I have only small remarks:

- _load_mask can be outside of the mask class (not dependent on class) and be renamed to read_mask_file (load implies for me, that it is put into some variable inside of a class, read would imply it is returned as value)
- all functions with the flipud parameter should have documentation, since it is not immediately obvious what the parameter means
- for read_mask_file the flipud parameter should probably also be default
- small note regarding styling: my linter wants to have spaces around "=" in the function declaration - do you use different styles?
- pyfixtures are perfect, i did not know about them when I started in 2013, but will write most of the tests now with fixtures

Thank you!

File: dioptas/controller/MaskController.py [hunk 7]
```diff
@@ -368,24 +368,35 @@ def cosmic_btn_click(self):
...
 
+    def _is_mask_flipud(self, filename: str) -> bool:
+        if filename.endswith('.npy') or filename.endswith('.edf'):
+            return self.model.img_model.img_flipud
+        return False
+
+    def _mask_filters(self) -> str:
+        if self.model.img_model.img_flipud:
+            return 'Mask (*.mask);;Vertically flipped mask (*.npy *.edf)'
+        return 'Mask (*.mask);;Mask (*.npy *.edf)'
```
[the remark this answers is not in the document]
Comment (CPrescher) [reply dioptas/controller/MaskController.py:379]: I am not sure I understand this logic, in the current implementation img_flipud is always false. I honestly would not try to track this, and always have the option to save normal mask or flipped masked.

[the remark this answers is not in the document]
Comment (t20100) [reply dioptas/controller/MaskController.py:379]: `img_flipud` is `True` for images loaded with PIL, fabio and lambda loaders and `False` for the others (spe, karabo, hdf5), so I handled this automatically.

But indeed, it's probably not worth it and can be more confusing than explicit options. I'll update that.

**Comment by t20100** [comment]
> _load_mask can be outside of the mask class (not dependent on class) and be renamed to read_mask_file

In which module should it go then? `dioptas.model.util.HelperModule`?

I didn't lint this PR. Which linter do you use?

**Comment by CPrescher** [comment]
I am also unsure where to put it, but looking into the class, there are already static methods already on the MaskModel class, maybe just make it static and leave it inside, otherwise probably also OK inside of the MaskModel module

I used default auto-formatting in PyCharm. However, looking into it, I found that for:

def function(default_param="blub"):

no space is used, but for:

def function(default_typed_param: bool = True):

the auto-formatting wants to use spaces. I guess it is a matter of preference:), both are fine for me

File: dioptas/controller/MaskController.py [hunk 8]
```diff
@@ -370,35 +375,46 @@ def cosmic_btn_click(self):
 
     def save_mask_btn_click(self):
         img_filename, _ = os.path.splitext(os.path.basename(self.model.img_model.filename))
-        filename = save_file_dialog(self.widget, "Save mask data",
-                                    os.path.join(self.model.working_directories['mask'],
-                                                 img_filename + '.mask'),
-                                    filter='Mask (*.mask)')
+        filename, selected_filter = QtWidgets.QFileDialog.getSaveFileName(
```
Comment (t20100) [inline dioptas/controller/MaskController.py:378]: I replaced `save_file_dialog` with a direct call to `QtWidgets.QFileDialog.getSaveFileName` to retrieve the selected filter. And I use this to toggle the vertical flip. This paves the way to support the same file extensions for both options.

From what I tested with PyQt5 and PyQt6, a 2-tuple is always returned and the call to `set_extension` is not needed

https://github.com/Dioptas/Dioptas/blob/772f6db87a7a4a929de7c90ca58c1c73f81ef6c1/dioptas/widgets/UtilityWidgets.py#L199-L205

File: dioptas/controller/MaskController.py [hunk 9]
```diff
@@ -370,35 +375,46 @@ def cosmic_btn_click(self):
...
         if filename != '':
+            flipud = selected_filter == self.FLIPUD_MASK_FILTER
             self.model.working_directories['mask'] = os.path.dirname(filename)
-            self.model.mask_model.save_mask(filename)
+            self.model.mask_model.save_mask(filename, flipud)
 
     def load_mask_btn_click(self):
-        filename = open_file_dialog(self.widget, caption="Load mask data",
-                                    directory=self.model.working_directories['mask'], filter='*.mask')
+        filename, selected_filter = QtWidgets.QFileDialog.getOpenFileName(
```
Comment (t20100) [inline dioptas/controller/MaskController.py:391]: Same here, `open_file_dialog` wrapper no longer seems to be needed.

**Comment by t20100** [comment]
I updated the PR to address the comments:

- Removed `img_flipud` logic 
- commit b9e71a00c618a1f4382dd02c3a217612e7b6fcc4 : Use 2 explicit options in the dialogs instead: `Mask` and `Vertically flipped mask`
- commit 406e49316c450ece556022a4679369dc9cc3eb0d :
  - Renamed `_load_mask` to `read_mask_file` and made static
  - documented `flipud` parameter
  - Set default to `read_mask_file`'s `flipud` parameter
  - Fixed typed argument linting: white space around `=` looks to be the standard.
- Updated tests to accommodate with the changes

File: dioptas/controller/MaskController.py [hunk 10]
```diff
@@ -33,6 +33,11 @@
 
 
 class MaskController(object):
+
+    DEFAULT_MASK_FILTER = 'Mask (*.mask)'
```
Comment (t20100) [inline dioptas/controller/MaskController.py:37]: I hesitate to add `.npy` and `.edf` also to the default:
```suggestion
    DEFAULT_MASK_FILTER = 'Mask (*.mask *.npy *.edf)'
```
The remaining of the code should work fine.

**Comment by CPrescher** [comment]
Thanks, i think it looks good. I think the current separation of defaulting to only see *.mask files upon loading is ok. For saving I am not sure whether *.npy and *.edf should be combined. This would currently mean you manually have to delete the npy ending and type in edf as a file ending, which is odd and an unusual user experience - for loading they could still be combined row though.

**Comment by t20100** [comment]
I updated the dialog options as you proposed:
- one filter per format for saving (indeed much better, thanks!)
- multiple format per option for loading.

Does something need to be changed for the batch?

File: dioptas/controller/MaskController.py [hunk 11]
```diff
@@ -370,35 +374,56 @@ def cosmic_btn_click(self):
...
-                                    os.path.join(self.model.working_directories['mask'],
-                                                 img_filename + '.mask'),
-                                    filter='Mask (*.mask)')
+        filename, selected_filter = QtWidgets.QFileDialog.getSaveFileName(
+            self.widget,
+            caption="Save mask data",
+            directory=os.path.join(self.model.working_directories['mask'],
+                                   img_filename + '.mask'),
+            filter=';;'.join([
+                self.DEFAULT_MASK_FILTER,
```
Comment (t20100) [inline dioptas/controller/MaskController.py:383]: For completeness, shouldn't we add?

```suggestion
            filter=';;'.join([
                self.DEFAULT_MASK_FILTER,
                "Mask (*.npy)",
                "Mask (*.edf)",
```

File: dioptas/controller/MaskController.py [hunk 12]
```diff
@@ -370,35 +374,56 @@ def cosmic_btn_click(self):
...
 
     def load_mask_btn_click(self):
-        filename = open_file_dialog(self.widget, caption="Load mask data",
-                                    directory=self.model.working_directories['mask'], filter='*.mask')
+        filename, selected_filter = QtWidgets.QFileDialog.getOpenFileName(
+            self.widget,
+            caption="Load mask data",
+            directory=self.model.working_directories['mask'],
+            filter=';;'.join([
+                self.DEFAULT_MASK_FILTER,
```
Comment (t20100) [inline dioptas/controller/MaskController.py:399]: Same here:
```suggestion
                self.DEFAULT_MASK_FILTER,
                "Mask (*.npy *.edf)",
```

File: dioptas/controller/MaskController.py [hunk 13]
```diff
@@ -370,35 +374,56 @@ def cosmic_btn_click(self):
...
 
     def add_mask_btn_click(self):
-        filename = open_file_dialog(self.widget, caption="Add mask data",
-                                    directory=self.model.working_directories['mask'], filter='*.mask')
+        filename, selected_filter = QtWidgets.QFileDialog.getOpenFileName(
+            self.widget,
+            caption="Add mask data",
+            directory=self.model.working_directories['mask'],
+            filter=';;'.join([
+                self.DEFAULT_MASK_FILTER,
```
Comment (t20100) [inline dioptas/controller/MaskController.py:419]:
```suggestion
                self.DEFAULT_MASK_FILTER,
                "Mask (*.npy *.edf)",
```

**Comment by CPrescher** [comment]
Thanks. If you do not have anything more, you want to add, I would merge the pull request. Is this ok with you?

**Comment by t20100** [comment]
It's fine for me as it is.
Thanks for the review!

**Pull request merged**